mlflow/metrics/genai/genai_metric.py

```diff
@@ -77,18 +85,151 @@ def _extract_score_and_justification(text):
+            score, justification = future.result()
+            scores[indx] = score
+            justifications[indx] = justification
+
+    return scores, justifications
+
+
+def _get_aggregate_results(scores, aggregations):
+    # loop over the aggregations and compute the aggregate results on the scores
+    def aggregate_function(aggregate_option, scores):
+        import numpy as np
+
+        options = {
+            "min": np.min,
+            "max": np.max,
+            "mean": np.mean,
+            "median": np.median,
+            "variance": np.var,
+            "p90": lambda x: np.percentile(x, 90) if x else None,
+        }
+
+        if aggregate_option not in options:
+            raise MlflowException(
+                message=f"Invalid aggregate option {aggregate_option}.",
+                error_code=INVALID_PARAMETER_VALUE,
+            )
+
+        return options[aggregate_option](scores)
+
+    scores_for_aggregation = [score for score in scores if score is not None]
+
+    return (
+        {option: aggregate_function(option, scores_for_aggregation) for option in aggregations}
+        if aggregations is not None
+        else {}
+    )
+
+
+def _make_custom_genai_metric(
+    name: str,
+    judge_prompt: Optional[str] = None,
+    model: Optional[str] = _get_default_model(),
+    parameters: Optional[Dict[str, Any]] = None,
+    aggregations: Optional[List[str]] = ["mean", "variance", "p90"],  # noqa: B006
+    greater_is_better: bool = True,
+    max_workers: int = 10,
+    metric_metadata: Optional[Dict[str, Any]] = None,
+) -> EvaluationMetric:
+    def eval_fn(
+        *args,
+        **kwargs,
+    ) -> MetricValue:
+        """
+        This is the function that is called when the metric is evaluated.
+        """
+        prompt_template = PromptTemplate([judge_prompt, _PROMPT_FORMATTING_WRAPPER])
+        missing_variables = prompt_template.variables - set(kwargs.keys())
+        if missing_variables:
+            raise MlflowException(
+                message=f"Missing variable inputs to eval_fn: {missing_variables}",
+                error_code=INVALID_PARAMETER_VALUE,
+            )
+        grading_payloads = pd.DataFrame(kwargs).to_dict(orient="records")
+        arg_strings = [prompt_template.format(**payload) for payload in grading_payloads]
+        scores, justifications = _score_model_on_payloads(
+            arg_strings, model, parameters, max_workers
+        )
+
+        aggregate_scores = _get_aggregate_results(scores, aggregations)
+
+        return MetricValue(scores, justifications, aggregate_scores)
+
+    return make_metric(
+        eval_fn=eval_fn,
+        greater_is_better=greater_is_better,
+        name=name,
+        metric_metadata=metric_metadata,
+    )
+
+
 @experimental
 def make_genai_metric(
     name: str,
-    definition: str,
-    grading_prompt: str,
+    definition: Optional[str] = None,
+    grading_prompt: Optional[str] = None,
+    judge_prompt: Optional[str] = None,
```

Comment:
.... and now I'm torn. 
I'm all for API consolidation for the most part, but I'm not 100% sold on the behavior change of "if I add arg x, then args y, z, and q will be ignored" behavior. It feels like it's a bit confusing for users. It makes perfect sense when reading the docstring, it will make sense in the docs, but in the trenches, writing applied code that uses this API, I just have this nagging suspicion that this might be frustrating. 

What are your thoughts on `make_custom_genai_metric(name: str, judge_prompt: str, version: Optional[str], model: Optional[str], greater_is_better: Optional[bool], max_workers: Optional[int], metric_metadata: Optional[Dict[str, Any]]):` ?

It really feels like two separate APIs to me. Thoughts?

Reply:
Makes perfect sense and I agree - I think the minor issue with the name `make_custom_genai_metric` is that we already _imply_ you can create a "custom genai metric" - that is by passing in a metric definition/grading prompt to `make_genai_metric`.
But this is a naming thing, because I agree it's essentially two different APIs. The code right now stuffs two (disparate) sets of parameters into one

Reply:
I think with a solid name and a good explanation in the docs (follow up PR :) ) this will be very clear about the distinction between the two. 
`make_gen_ai_metric` is "custom with those kiddie inflatable bumpers that they put up in bowling alleys so your kids don't rage at how hard bowling is", while `apurva_gen_ai_metric` is "choose your own adventure, human. Instruct me how I shall do thy bidding"

Reply:
Sounds good to me, since this is time sensitive I'm happy with `make_custom_genai_metric` for the release since if needed we can clarify later (with some leeway given this is `@experimental`).

Reply:
Done. leaving this discussion open for visibility